autosklearn/ensembles/multi_objective_wrapper.py

```diff
@@ -0,0 +1,140 @@
+        self,
+        task_type: int,
+        metrics: Sequence[Scorer] | Scorer,
+        ensemble_class: Type[AbstractEnsemble],
+        ensemble_kwargs: Dict,
+        random_state: int | np.random.RandomState | None,
+        n_weights: int = 100,
+    ) -> None:
+        self.task_type = task_type
+        self.metrics = metrics if isinstance(metrics, Sequence) else [metrics]
+        self.ensemble_class = ensemble_class
+        self.ensemble_kwargs = ensemble_kwargs
+        self.n_weights = n_weights
+        self.random_state = check_random_state(random_state)
+
+    def __getstate__(self) -> Dict[str, Any]:
+        # Cannot serialize a metric if
+        # it is user defined.
+        # That is, if doing pickle dump
+        # the metric won't be the same as the
+        # one in __main__. we don't use the metric
+        # in the EnsembleSelection so this should
+        # be fine
+        self.metrics = None  # type: ignore
+        return self.__dict__
+
+    def fit(
+        self,
+        base_models_predictions: List[np.ndarray] | np.ndarray,
+        true_targets: np.ndarray,
+        model_identifiers: List[Tuple[int, int, float]],
+    ) -> AbstractEnsemble:
+        ensembles = []
+        losses = []
+        all_weights = self.random_state.rand(self.n_weights - 2, len(self.metrics))
+        all_weights = all_weights / np.sum(all_weights, axis=1).reshape((-1, 1))
+        all_weights = [[1.0, 0.0]] + list(all_weights) + [[0.0, 1.0]]
+        all_weights = sorted(all_weights, key=lambda x: x[0], reverse=True)
+        for weights in all_weights:
+            metric = WeightedMetricScorer(metrics=self.metrics, weights=weights)
+
+            ensemble = self.ensemble_class(
+                task_type=self.task_type, metrics=metric, **self.ensemble_kwargs
+            )
+            ensemble.fit(
+                base_models_predictions=base_models_predictions,
+                true_targets=true_targets,
+                model_identifiers=model_identifiers,
+            )
+            prediction = ensemble.predict(
+                base_models_predictions=base_models_predictions
+            )
+            losses.append(
+                calculate_losses(
+                    solution=true_targets,
+                    prediction=prediction,
+                    task_type=self.task_type,
+                    metrics=self.metrics,
+                )
+            )
+            ensembles.append(ensemble)
+
+        # Prune to the Pareto front!
+        losses_array = np.array(
+            [[losses_[metric.name] for metric in self.metrics] for losses_ in losses]
+        )
+        pareto_front = is_pareto_efficient(losses_array)
+        ensembles = [ensembles[i] for i, pareto in enumerate(pareto_front) if pareto]
+        self.ensembles_ = ensembles
+        return self
+
+    def predict(self, predictions: Union[np.ndarray, List[np.ndarray]]) -> np.ndarray:
+        return self.ensembles_[0].predict(predictions)
```

Comment:
I guess this means we're just predicting with the ensemble that did best with the weights `[1.0, 0.0]` which translates to the ensemble that did best with the first metric?

Would it make sense to average all predictions instead. It would be linearly more costly and maybe it doesn't make sense at all though.

Reply:
Yes, this means predicting with the first ensemble. However, it could also be a different weight if such an ensemble produced a better score for the first metric than the ensemble trained with the weights `[1.0, 0.0]`. Averaging all predictions wouldn't make sense to me because we then get unpredictable behavior again (we don't know what the predictions were optimized for...).